src/hotspot/share/gc/shared/slidingForwarding.cpp

```diff
@@ -0,0 +1,162 @@
+  size_t heap_size_words = heap.end() - heap.start();
+  if (UseSerialGC && heap_size_words <= (1 << NUM_COMPRESSED_BITS)) {
+    // In this case we can treat the whole heap as a single region and
+    // make the encoding very simple.
+    _num_regions = 1;
+    _region_size_words = round_up_power_of_2(heap_size_words);
+    _region_size_words_shift = log2i_exact(_region_size_words);
+  }
+}
+
+SlidingForwarding::~SlidingForwarding() {
+  if (_bases_table != nullptr) {
+    FREE_C_HEAP_ARRAY(HeapWord*, _bases_table);
+    _bases_table = nullptr;
+  }
+  if (_fallback_table != nullptr) {
+    delete _fallback_table;
+    _fallback_table = nullptr;
+  }
+}
+
+void SlidingForwarding::begin() {
+  assert(_bases_table == nullptr, "Should be uninitialized");
+  size_t max = _num_regions * NUM_TARGET_REGIONS;
+  _bases_table = NEW_C_HEAP_ARRAY(HeapWord*, max, mtGC);
+  for (size_t i = 0; i < max; i++) {
+    _bases_table[i] = UNUSED_BASE;
+  }
+}
+
+void SlidingForwarding::end() {
+  assert(_bases_table != nullptr, "Should be initialized");
+  FREE_C_HEAP_ARRAY(HeapWord*, _bases_table);
+  _bases_table = nullptr;
+
+  if (_fallback_table != nullptr) {
+    delete _fallback_table;
+    _fallback_table = nullptr;
+  }
+}
+
+void SlidingForwarding::fallback_forward_to(HeapWord* from, HeapWord* to) {
+  if (_fallback_table == nullptr) {
+    _fallback_table = new FallbackTable();
+  }
+  _fallback_table->forward_to(from, to);
+}
+
+HeapWord* SlidingForwarding::fallback_forwardee(HeapWord* from) const {
+  assert(_fallback_table != nullptr, "fallback table must be present");
+  return _fallback_table->forwardee(from);
+}
+
+FallbackTable::FallbackTable() {
+  for (size_t i = 0; i < TABLE_SIZE; i++) {
+    _table[i]._next = nullptr;
+    _table[i]._from = nullptr;
+    _table[i]._to   = nullptr;
+  }
+}
+
+FallbackTable::~FallbackTable() {
+  for (size_t i = 0; i < TABLE_SIZE; i++) {
+    FallbackTableEntry* entry = _table[i]._next;
+    while (entry != nullptr) {
+      FallbackTableEntry* next = entry->_next;
+      FREE_C_HEAP_OBJ(entry);
+      entry = next;
+    }
+  }
+}
+
+size_t FallbackTable::home_index(HeapWord* from) {
+  uint64_t val = reinterpret_cast<uint64_t>(from);
+  // This is the mixer stage of the murmur3 hashing:
+  // https://github.com/aappleby/smhasher/blob/master/src/MurmurHash3.cpp
+  val ^= val >> 33;
+  val *= 0xff51afd7ed558ccdULL;
+  val ^= val >> 33;
+  val *= 0xc4ceb9fe1a85ec53ULL;
+  val ^= val >> 33;
+  // Shift to table-size.
+  val = val >> (64 - log2i_exact(TABLE_SIZE));
+  size_t idx = static_cast<size_t>(val);
+  assert(idx < TABLE_SIZE, "must fit in table: idx: " SIZE_FORMAT ", table-size: " SIZE_FORMAT ", table-size-bits: %d",
+         idx, TABLE_SIZE, log2i_exact(TABLE_SIZE));
+  return idx;
+}
+
+void FallbackTable::forward_to(HeapWord* from, HeapWord* to) {
+  size_t idx = home_index(from);
+  if (_table[idx]._from != nullptr) {
```

Comment:
Here you need to do a contains check, right? Because, as you wrote in your answer to Aleksey, forwardings can be rewritten: https://github.com/openjdk/jdk/pull/13582/files#r1180126262

Reply:
I don't think that ever happens (I think we'd only ever re-forward from normal forwarding to fallback-forwarding once), but I am adding that check for extra sanity.